### DISCUSSION TITLE
Localize with different camera

### DISCUSSION THREAD

**Discussion by rebellojason7** [stella-cv/stella_vslam#130] [Q&A] [answered]
Hi,

I built a map using a stereo camera where the image size was 1280 by 1024 and I am trying to localize a different camera in the map with image size 1344 by 720. However I can't seem to localize the new camera. I have calibrated the new camera and fed in the right parameters. I looked at this issue https://github.com/xdspacelab/openvslam/issues/388. but cant seem to get my data to work. Any thoughts ? I can share my data if need be.
Thanks

**Comment by ymd-stella** [comment, marked as answer]
@rebellojason7 
Please align the resolution per angle of view. Is only one of your monocular or stereo cameras wide FoV?

If you can identify which step is failing you, we may be able to provide additional advice.

**Reply by rebellojason7** [reply]
Hi @ymd-stella , Thanks for the reply. I'm not quite sure what you mean by "align the resolution per angle of view".

So I built a map using a stereo camera (consisting of two individual cameras). and then localize using a different camera (not the ones used to build the map). I have pasted the stereo and monocular configurations down below. I dont think there is any step that is failing (as in getting an error message), my monocular camera just doesn't localize. Any help in debugging this would be useful. Thanks !!!

#==============#
# Camera Model #
#==============#

# intrinsic parameters after stereo-rectification are written

Camera.name: "Ximea stereo"
Camera.setup: "stereo"
Camera.model: "perspective"

# new "rectified" matrices is the first three cols of the projection matrix which is calculated with cv::stereoRectify()
# e.g. fx = P1[0][0] or P2[0][0], cx = P1[0][2] or P2[0][2]
#      fy = P1[1][1] or P2[1][1], cy = P1[1][2] or P2[1][2]

Camera.fx: 639.874464
Camera.fy: 639.874464
Camera.cx: 685.940071
Camera.cy: 505.437786

# there is no distortion after stereo rectification

Camera.k1: 0.0
Camera.k2: 0.0
Camera.p1: 0.0
Camera.p2: 0.0
Camera.k3: 0.0

# focal_x_baseline is -P2[0][3] which is calculated with cv::stereoRectify()

Camera.fps: 20.0
Camera.cols: 1280
Camera.rows: 1024
Camera.focal_x_baseline: 120.561564

Camera.color_order: "Gray"

#======================#
# Stereo Rectification #
#======================#

# original intrinsic parameters (K, D) and stereo-recitification parameters (R)
# matrices (K, R) are written in row-major order

StereoRectifier.K_left: [664.222256, 0.000000, 637.477721, 0.000000, 664.253392, 499.682345, 0.000000, 0.000000, 1.000000]
StereoRectifier.D_left: [-0.184252, 0.053314, -0.000235, 0.000791, 0.000000]
StereoRectifier.R_left: [0.998635, -0.034469, -0.039235, 0.033891, 0.999308, -0.015321, 0.039736, 0.013970, 0.999113]
StereoRectifier.K_right: [663.366227, 0.000000, 650.615117, 0.000000, 662.833368, 504.218877, 0.000000, 0.000000, 1.000000]
StereoRectifier.D_right: [-0.185886, 0.053334, 0.000421, -0.000440, 0.000000]
StereoRectifier.R_right: [0.998096, -0.048540, -0.038051, 0.049092, 0.998700, 0.013722, 0.037336, -0.015564, 0.999182]

#================#
# ORB Parameters #
#================#

Feature.max_num_keypoints: 1000
Feature.scale_factor: 1.2
Feature.num_levels: 8
Feature.ini_fast_threshold: 20
Feature.min_fast_threshold: 7

#========================#
# Initializer Parameters #
#========================#

Initializer.num_min_triangulated_pts: 100

#===========================#
# PangolinViewer Parameters #
#===========================#

PangolinViewer.keyframe_size: 0.07
PangolinViewer.keyframe_line_width: 1
PangolinViewer.graph_line_width: 1
PangolinViewer.point_size: 2
PangolinViewer.camera_size: 0.08
PangolinViewer.camera_line_width: 3
PangolinViewer.viewpoint_x: 0
PangolinViewer.viewpoint_y: -0.65
PangolinViewer.viewpoint_z: -1.9
PangolinViewer.viewpoint_f: 400

and then I localize using a monocular camera and this is my config file:

%YAML:1.0

#--------------------------------------------------------------------------------------------
# Camera Parameters. Adjust them!
#--------------------------------------------------------------------------------------------

# Camera calibration and distortion parameters (OpenCV) 
Camera.name: "ros"
Camera.setup: "monocular"
Camera.model: "perspective"

Camera.fx: 1162.321800
Camera.fy: 1176.453533
Camera.cx: 684.0316951
Camera.cy: 351.7395287

Camera.k1: -0.0041610
Camera.k2: -0.0081600
Camera.p1: -0.0008450
Camera.p2: 0.0042193
Camera.k3: 0

# Camera frames per second 
Camera.fps: 30
Camera.cols: 1344
Camera.rows: 720

Camera.color_order: "Gray"

#================#
# ORB Parameters #
#================#

Feature.max_num_keypoints: 2000
Feature.scale_factor: 1.2
Feature.num_levels: 8
Feature.ini_fast_threshold: 20
Feature.min_fast_threshold: 7

#===========================#
# PangolinViewer Parameters #
#===========================#

PangolinViewer.keyframe_size: 0.06
PangolinViewer.keyframe_line_width: 1
PangolinViewer.graph_line_width: 1
PangolinViewer.point_size: 2
PangolinViewer.camera_size: 0.07
PangolinViewer.camera_line_width: 2
PangolinViewer.viewpoint_x: 0
PangolinViewer.viewpoint_y: -20
PangolinViewer.viewpoint_z: -0.1
PangolinViewer.viewpoint_f: 2800

**Reply by ymd-stella** [reply]
> I'm not quite sure what you mean by "align the resolution per angle of view".

If I'm not wrong in my calculations, the horizontal angle of view of the stereo camera is 90 degrees and the horizontal angle of view of the monocular camera is 60 degrees. (It is calculated from the resolution and focal length.)

When the horizontal resolution of the stereo camera with a 90-degree angle of view is 1280, the recommended horizontal resolution for the camera with a 60-degree angle of view is 853. (60. / 90. * 1280= 853)

**Reply by ymd-stella** [reply]
In addition, the monocular camera you have prepared is not be suitable for Visual SLAM. Is it possible to prepare a camera with a horizontal angle of view of 90 degrees or more?

**Reply by rebellojason7** [reply]
I see. I'm not too sure about the values and I dont think I can modify my image values. I am able to build a brand new map with the monocular camera that I try to localize, which means that Openvslam can handle that sort of image form the monocular camera. The problem occurs when I build a map with the stereo pair and try to localize with the monocular camera in the stereo map. I can attach some pictures and a video tomorrow, but in general I dont know why the monocular camera wont localize

**Reply by ymd-stella** [reply]
If you are not going to change the resolution or the angle of view, there is nothing more I can advise. Good luck!

**Reply by surfii3z** [reply]
Hi, I have tried doing similar things. I will share my experience.

The following videos I tried to make the map with RGB-D camera Realsense-D435i (scale-aware) and localize with DJI Tello camera (see the spec here https://www.ryzerobotics.com/tello/specs)

https://youtu.be/i1J0tmgbCUc (<<< this was ORB_SLAM2)

https://youtu.be/cx2MV2OOG7U

In both case, all of the video are rectified before feeding into the openvslam.

My understanding is that as long as your features are repeatable and the map is not corrupted, the map points should be projected to reconstruct the real 3D structure. Therefore, it should be working.

**Reply by surfii3z** [reply]
I also did the experiment using the similar mapping (D455 and D435i) and using DJI tello for localization.

You could see the videos here

https://docs.google.com/document/d/1AUxfD6Pu9NnbsJWb-hfafaDE983TbioAylu3eSk3vbI/edit?usp=sharing.

In this experiment, the indoor environment is quite large so the D435i could not provide reliable depth (also due to reflective wall) therefore, it made a corrupted map and I could not even localize with the same camera.

I think it is hard to debug if we don't know the environment you tried to do the mapping.

However, I think you could try two things

1) Try to do stereo mapping with the stereo camera and try to localize with one of your side stereo camera.

2) Try to do the mapping with monocular mode (one side of your stereo cam or that mono cam) and relocalize with the monocular mode (one side of your stereo cam or that mono cam).

This might help us to debug better.

**Reply by ymd-stella** [reply]
If it is failing in the image search step, I recommend that you try increasing the number of features to be extracted. (`Feature.max_num_keypoints`)

**Reply by rebellojason7** [reply]
Hi @ymd-stella and @surfii3z. First of all, thanks again for offering to help. Let me state the problem. I have 3 cameras: a stereo pair and a monocular camera. The image sizes are mentioned below. I managed to change the monocular size from the above mentioned size but only a little.
Monocular camera (camera 1) size 1280 x 720
Stereo camera (camera 2 and camera 3) are each of size 1280 x 1024.

I also created 2 files in ros. Feel free to use them. I can also attach the launch files if you need. Please note that the online slam version, you will have to modify the topics in openvslam_ros.cc:
1. a file that allows for online slam or localization 
2. a file for offline slam and localization.

[online_slam.txt](https://github.com/OpenVSLAM-Community/openvslam/files/6736349/online_slam.txt)

[offline_slam.txt](https://github.com/OpenVSLAM-Community/openvslam/files/6736351/offline_slam.txt)

Problem 1. Using the online_slam version of my code. I can build a map using each of the 3 cameras individually as well as the stereo version (using camera2 and camera3) in stereo config. This proves that each camera is good for slam (in my opinion). Please note that when I build the stereo map, I use the -r option which rectifies the images ( i think). The problem comes when I try to localize each of the cameras individually in the stereo map. I can localize camera2 and camera3 individually, however my monocular camera (camera 1) doesnt localize in the stereo map. Should I rectify camera1 images before trying to localize ? I dont rectify camera2 or camera3 in monocular mode since I thought that the distortion parameters are included in the projection.

Situation 2. Lets say I build 3 monocular maps using camera1, camera2 and camera3. Lets call them map_camera1, map_camera2, map_camera3. I find that the best features are obtained in my map_camera1 and when I try to localize camera2 and camera3 (in mono mode) in my map_camera1, it works. However when I try to localize camera1 in map_camera2 or map_camera3 it doesnt work too well.

Thoughts: The main problem I am facing is that I need a map with scale, therefore I use the stereo camera to create the map. I then need to localize each of the 3 cameras individually in the stereo map. camera2 and camera3 localize well (makes sense since they were used in stereo mode to build the map) However camera1 doesnt localize at all. Any thoughts ? I would be open to chatting on slack if that would be better. Thanks for the help !!!

**Reply by surfii3z** [reply]
Thanks for the details debugging experiment. Could you show us a little bit of your experiment? Like a short video or snapshots.

What do you mean by best features are obtained in our map_camera1, what happens to the map in cam2 and cam3??

How about making the map in Stereo mode using cam2, cam3 and do localization in monocular mode using cam2, cam3, does it work? (cam2 mapping (stereo) -> cam2, cam3 mapping (mono))  (cam3 mapping (stereo) -> cam2, cam3 mapping (mono))

**Reply by surfii3z** [reply]
Otherwise you might try to make a denser map with stereo configuration with more variety of the angles such that the mono mode of the first cam could determine `3d-to-2d correspondence` easier.

**Reply by libertyfree** [reply]
I built a map using a perspective camera  ,location use monocular can not location.this  stella_vslam lib is not suuport or my config is not correct?
